designs/2021-stricter-rule-test-validation/README.md

```diff
@@ -0,0 +1,286 @@
+- Repo: eslint/eslint
+- Start Date: 2021-11-30
+- RFC PR: <https://github.com/eslint/rfcs/pull/84>
+- Authors: [bmish](https://github.com/bmish)
+
+# Stricter validation of rule tests
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+We will add additional assertions to the rule test runner to address several potential points of poor test coverage, especially related to [rule suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions).
+
+Assertions added to invalid test case error objects:
+
+- Must contain `suggestions` if the test case produces suggestions
+- Must contain the violation message (`message` or `messageId`, but not both)
```

Comment:
How to make sure message is generated by `messageId` and `data`(not calling `content.report()` with `message` property)?

How to make sure the template string is replaced correctly?

Reply:
You can use `messageId` and `data` in test cases: https://eslint.org/docs/developer-guide/nodejs-api#testing-errors-with-messageid

> How to make sure message is generated by `messageId` and `data`(not calling `content.report()` with `message` property)?

If you specify `messageId` in the test case, it asserts that `context.report()` was called with a `messageId`, and checks if that `messageId` is the same as the `messageId` specified in the test case.

> How to make sure the template string is replaced correctly?

If you also specify `data` in the test case, RuleTester will get the meta message (by `messageId` specified in the test case) from `rule.meta.messages`, and replace placeholders with `data` from the test case to calculate the expected message. Then, it asserts that the expected message is the same as the reported `message` (which was calculated from `messageId` and `data` passed to `context.report()`).

Assertions that use `messageId` and `data` from the test case are here: https://github.com/eslint/eslint/blob/18f5e05bce10503186989d81ca484abb185a2c9d/lib/rule-tester/rule-tester.js#L814-L843

Reply:
I know how to do this test for current version. But this line forbid test both `messageId` and `message`.I'll have to do this test separately.

Reply:
Why should we forbid assert `{message, messageId, data}` together? So ONE test can make sure the message is generated by `messageId` and `data`, the template is also replaced successfully.

Reply:
This question seems tangential to the RFC. Can we open a separate issue to discuss?

Reply:
Not really interested. I barely use the builtin tester.

Reply:
Okay, then I’ll mark this discussion as resolved.

Reply:
Sorry to reopen this, but I would like to ask some clarifications.

There's an existing rule tester assertion today `Error should not specify both 'message' and a 'messageId'`. The only thing I'm changing here is that you have to specify the message in SOME form, as previously it was possibly to omit any mention of the message.

So @fisker, is it just this existing assertion you are unhappy with?

Do let me know if you have a problem with new behavior in the RFC so we can consider your use cases.

Reply:
I haven't used the built-in tester for a while, the first time I want test both `messageId` + `data` and `message` was we first use `suggestion` api, I thought `data` was shared between error and suggestions at first, but found the message template was not replaced correctly.

See https://github.com/sindresorhus/eslint-plugin-unicorn/pull/635

Since then, I always add an extra test to test the actual message/desc https://github.com/sindresorhus/eslint-plugin-unicorn/blob/c501243e7f9fb012b09a69ffc4ab66fb3376bb7d/test/no-array-callback-reference.mjs#L286

Reply:
I closed this because it’s not related to the RFC specifically and shouldn’t distract from the larger discussion. 

Please move the discussion to an issue if you’d like to continue it.